In [ ]:
%%html
<style>

/* ── Variables ── */
:root {
    --font-sans: sans-serif;
    --font-mono: monospace;

    --font-size: 1.17rem;
    --code-letter-spacing: -0.03rem;
    --font-weight: 400;
    --quote-weight: 300;
    --border-thin: 1px;
    --border-card: 1.5px;
    --border-thick: 7px;

    --color-foreground: hsla(0, 0%, 0%, 1.00);
    --color-muted-foreground: hsla(0, 0%, 0%, 0.7);

    --color-background: hsla(48, 21%, 89%, 1.00);
    --color-muted-background: hsla(46, 21%, 86%, 1.00);
    
    --muted-foreground: hsla(0, 0%, 0%, 0.35);
    --mute-background: hsla(0, 0%, 0%, 0.12);
    
    --color-title: hsla(0, 0%, 0%, 1.00);
    --color-accent1: hsla(47, 11%, 33%, 1.00);
    --color-accent2: hsla(162, 100%, 27%, 1.00);
    --color-accent3: hsla(248, 100%, 42%, 1.00);
    --color-accent4:hsla(338, 100%, 62%, 1.00);
    --color-accent5: hsla(230, 52%, 42%, 1.00);

    --color-cell-prompt: hsla(0, 0%, 0%, 1.00);
    --color-cell-raw: hsla(0, 0%, 0%, 1.00);
    --color-cell-note: hsla(145, 63%, 49%, 0);
    --color-cell-code: hsla(212, 72%, 59%, 0);
    --color-code: var(--color-accent5);
}

:root.dark {
    --color-foreground: hsla(0, 0%, 100%, 1.00);
    --color-muted-foreground: hsla(0, 100%, 100%, 0.45);
    
    --color-background:       hsla(0, 0%, 00%, 1);
    --color-muted-background: hsla(0, 0%, 12%, 1);
    
    --muted-foreground: hsla(0, 0%, 100%, 0.35);
    --mute-background: hsla(0, 0%, 100%, 0.3);
    
    --color-title: hsla(48, 22%, 77%, 1.00);
    --color-accent1: hsla(48, 42%, 82%, 1.00);
    --color-accent2: hsla(162, 100%, 50%, 1.00);
    --color-accent3: hsla(248, 100%, 82%, 1.00);
    --color-accent4:hsla(338, 100%, 62%, 1.00);
    --color-accent5: hsla(230, 100%, 80%, 1.00);
    --color-muted-accent1: hsla(49, 41%, 82%, 0.5);

    --color-cell-prompt: var(--color-accent1);
    --color-cell-raw: var(--color-accent1);
    --color-cell-note: hsla(145, 63%, 49%, 0);
    --color-cell-code: hsla(212, 72%, 59%, 0);
    --color-code: hsla(149, 100%, 84%, 1.00);
    --color-code: var(--color-accent5);

}

body {
    font-family: var(--font-sans);
    font-size: var(--font-size);
    font-weight: var(--font-weight);
}

#dialog-container {
    border-color: var(--color-background);
}

div.msg-card {
    border-left-width: var(--border-thin) !important;
}

pre {
    padding: 0.3rem 1rem !important;
    font-weight: 400;
}

.msg-content pre {
    background: var(--color-muted-background);
}

code {
    color: var(--color-foreground);
    font-family: var(--font-mono);
    font-size: calc( var(--font-size) * 0.93 );
}

.msg-content p code, .msg-content li code, .msg-content table code { background: transparent; }
:is(h1) code { font-size: inherit; font-weight: 800; }
:is(h2) code { font-size: inherit; font-weight: 700; }
:is(h3) code { font-size: inherit; }
:is(h4, h5, h6) code { font-size: inherit; }

.prose :not(pre, h1, h2, h3, h4, h5, h6)>code {
    padding: 0;
    font-size: var(--font-size);
    font-weight: var(--font-weight);
    letter-spacing: var(--code-letter-spacing);
}

.prose :not(pre)>code {
    color: var(--color-code);
    background: transparent;
    padding: 0;
}

.prose h1,h2,h3,h4,h5,h6 {
    color: var(--color-title);
    border-color: var(--color-title);
    border-style: solid;
}

.prose h1 {
    font-size: 3rem;
    text-align: center;
    margin: 20px 0 0px;
    padding: 12px 0;
    border-width: var(--border-thick) 0px var(--border-thick) 0px;
}

.prose h2 {
    font-size: 2.25rem;
    border-width: 0px 0px var(--border-thin) 0px;
}

.prose h3 {
    font-size: 1.5rem;
}

.prose h4 {
    font-size: 1.25rem;
}

.prose h5 {
    font-size: 1.1rem;
}

.prose h6 {
    font-size: 1rem;
}

#dialog-container {
    background: transparent;
}

.uk-card {
    background: var(--color-background);
    border-color: transparent;
    box-shadow: transparent;
}

.uk-card-header p {
    color: var(--muted-foreground) !important;
}

div[data-sm-header="collapsed"] .uk-card-header p {
    color: var(--color-accent4) !important;
}

nav .uk-btn {
    color: var(--muted-foreground);
}

nav .uk-btn-group {
    padding-right: 25px;
}

nav .uk-btn-group button {
    color: var(--color-accent1);
}

nav button[title="Expand heading"].uk-btn {
    color: var(--color-accent4);
}

div[data-sm-header="collapsed"] * {
    color: var(--color-muted-foreground);
}

div[data-icons-vis] button {
    color: var(--color-background);
}

div[data-icons-vis] .uk-btn-group:hover button {
    color: var(--color-accent1);
}

button[title="Toggle AI visibility (h)"] {
    position: absolute;
    right: 0;
}

.token-usage-details {
    color: var(--muted-foreground);
}

.msg-body {
    padding: var(--spacing) calc(var(--spacing) * 2.6);
}

.msg-card {
    background: var(--color-background);
    box-shadow: None;
}

.prose p {
    color: var(--color-foreground);
    font-size: var(--font-size);
}

.prose li {
    color: var(--color-foreground);
    font-size: var(--font-size);
}

.prose 
ol>li::marker,
ul>li::marker {
    color: var(--color-accent1);
}

.prose table {
    font-size: var(--font-size);
}

.prose thead {    border-bottom: var(--border-thin) solid var(--color-accent1); }
.prose thead th { color: var(--color-accent1); }
.prose tbody tr { border-bottom: var(--border-thin) solid var(--color-muted-foreground); }

.prose blockquote {
    font-style: normal;
    font-weight: var(--quote-weight);
    margin: 0;
    padding: 0.1rem 1rem;
    border-width: 0px 0px 0px 1px;
    border-style: solid;
    border-color: var(--color-accent1);
}

.prose h1+blockquote {
    border-width: 0;
}

.prose blockquote * {
    color: var(--color-accent1);
}

div.monaco-editor-background { background: var(--color-muted-background); }
div.margin { background: var(--color-muted-background) !important; }
div.view-line { font-family: var(--font-mono); }

.hl-constructor, .hl-function, .hl-function-method {
    color: var(--color-accent2);
}

.bg-muted {
    color: var(--muted-foreground);
    background: inherit;
}

.msg-card[data-mtype="prompt"] { 
    border-left-color: var(--color-cell-prompt) !important;
 }

.msg-card[data-mtype="raw"] { 
    border-left-color: var(--color-cell-raw) !important;
    border-left-width: var(--border-thick) !important;
 }

.msg-card[data-mtype="note"] { 
    border-left-color: var(--color-cell-note) !important;
 }

.msg-card[data-mtype="code"] { 
    border-left-color: var(--color-cell-code) !important;
 }

.editable[data-sm=primary]
{
    --tw-ring-color: var(--color-muted-accent1) !important;
    border-radius: 0px;
    
}
.editable[data-editing] {
    outline-color: var(--color-accent4);
}

</style>

HTML(<style>

/* ── Variables ── */
:root {
    --font-sans: sans-serif;
    --font-mono: monospace;

    --font-size: 1.17rem;
    --code-letter-spacing: -0.03rem;
    --font-weight: 400;
    --quote-weight: 300;
    --border-thin: 1px;
    --border-card: 1.5px;
    --border-thick: 7px;

    --color-foreground: hsla(0, 0%, 0%, 1.00);
    --color-muted-foreground: hsla(0, 0%, 0%, 0.7);

    --color-background: hsla(48, 21%, 89%, 1.00);
    --color-muted-background: hsla(46, 21%, 86%, 1.00);

    --muted-foreground: hsla(0, 0%, 0%, 0.35);
    --mute-background: hsla(0, 0%, 0%, 0.12);

    --color-title: hsla(0, 0%, 0%, 1.00);
    --color-accent1: hsla(47, 11%, 33%, 1.00);
    --color-accent2: hsla(162, 100%, 27%, 1.00);
    --color-accent3: hsla(248, 100%, 42%, 1.00);
    --color-accent4:hsla(338, 100%, 62%, 1.00);
    --color-accent5: hsla(230, 52%, 42%, 1.00);

    --color-cell-prompt: hsla(0, 0%, 0%, 1.00);
    --color-cell-raw: hsla(0, 0%, 0%, 1.00);
    --color-cell-note: hsla(145, 63%, 49%, 0);
    --color-cell-code: hsla(212, 72%, 59%, 0);
    --color-code: var(--color-accent5);
}

:root.dark {
    --color-foreground: hsla(0, 0%, 100%, 1.00);
    --color-muted-foreground: hsla(0, 100%, 100%, 0.45);

    --color-background:       hsla(0, 0%, 00%, 1);
    --color-muted-background: hsla(0, 0%, 12%, 1);

    --muted-foreground: hsla(0, 0%, 100%, 0.35);
    --mute-background: hsla(0, 0%, 100%, 0.3);

    --color-title: hsla(48, 22%, 77%, 1.00);
    --color-accent1: hsla(48, 42%, 82%, 1.00);
    --color-accent2: hsla(162, 100%, 50%, 1.00);
    --color-accent3: hsla(248, 100%, 82%, 1.00);
    --color-accent4:hsla(338, 100%, 62%, 1.00);
    --color-accent5: hsla(230, 100%, 80%, 1.00);
    --color-muted-accent1: hsla(49, 41%, 82%, 0.5);

    --color-cell-prompt: var(--color-accent1);
    --color-cell-raw: var(--color-accent1);
    --color-cell-note: hsla(145, 63%, 49%, 0);
    --color-cell-code: hsla(212, 72%, 59%, 0);
    --color-code: hsla(149, 100%, 84%, 1.00);
    --color-code: var(--color-accent5);

}

body {
    font-family: var(--font-sans);
    font-size: var(--font-size);
    font-weight: var(--font-weight);
}

#dialog-container {
    border-color: var(--color-background);
}

div.msg-card {
    border-left-width: var(--border-thin) !important;
}

pre {
    padding: 0.3rem 1rem !important;
    font-weight: 400;
}

.msg-content pre {
    background: var(--color-muted-background);
}

code {
    color: var(--color-foreground);
    font-family: var(--font-mono);
    font-size: calc( var(--font-size) * 0.93 );
}

.msg-content p code, .msg-content li code, .msg-content table code { background: transparent; }
:is(h1) code { font-size: inherit; font-weight: 800; }
:is(h2) code { font-size: inherit; font-weight: 700; }
:is(h3) code { font-size: inherit; }
:is(h4, h5, h6) code { font-size: inherit; }

.prose :not(pre, h1, h2, h3, h4, h5, h6)>code {
    padding: 0;
    font-size: var(--font-size);
    font-weight: var(--font-weight);
    letter-spacing: var(--code-letter-spacing);
}

.prose :not(pre)>code {
    color: var(--color-code);
    background: transparent;
    padding: 0;
}

.prose h1,h2,h3,h4,h5,h6 {
    color: var(--color-title);
    border-color: var(--color-title);
    border-style: solid;
}

.prose h1 {
    font-size: 3rem;
    text-align: center;
    margin: 20px 0 0px;
    padding: 12px 0;
    border-width: var(--border-thick) 0px var(--border-thick) 0px;
}

.prose h2 {
    font-size: 2.25rem;
    border-width: 0px 0px var(--border-thin) 0px;
}

.prose h3 {
    font-size: 1.5rem;
}

.prose h4 {
    font-size: 1.25rem;
}

.prose h5 {
    font-size: 1.1rem;
}

.prose h6 {
    font-size: 1rem;
}

#dialog-container {
    background: transparent;
}

.uk-card {
    background: var(--color-background);
    border-color: transparent;
    box-shadow: transparent;
}

.uk-card-header p {
    color: var(--muted-foreground) !important;
}

div[data-sm-header="collapsed"] .uk-card-header p {
    color: var(--color-accent4) !importa

In [ ]:
from IPython.display import Markdown

# [1iis/srvbox](https://github.com/1iis/srvbox)<br>Server Baseline CRAFT

> A small production server setup, in Python.

`srvbox` is a minimal, hardened, boring production server baseline.  
If [`devbox`](https://github.com/1iis/devbox) is the comfortable SWE workstation, `srvbox` is the lean production substrate.

It provisions the host into a known baseline, then gets out of the way.

## Current status

`srvbox` has reached its first MVP: a small, artifact-deployed, hardened production server baseline for Ubuntu/Debian-like hosts.

The deployment spine is:

```text
local repo
→ scripts/deploy.sh
→ tar artifact + generated runner
→ scp to /tmp/deploy/1iis/
→ ssh -t user@host "sudo sh runner"
→ scripts/setup.sh
→ /opt/1iis/srvbox/releases/<timestamp>/
→ /opt/1iis/srvbox/current
→ host/sync.py apply
```

The baseline now includes:

- root-owned timestamped releases under `/opt/1iis/srvbox/releases/`
- read-only `status` and root-only `apply` reconciler commands
- base `/opt/1iis`, `/etc/1iis`, `/var/lib/1iis`, `/var/log/1iis` directories
- explicit access policy: human users unmanaged; deploy user pre-exists with SSH + sudo
- SSH hardening via validated `/etc/ssh/sshd_config.d/` drop-in
- `ufw` firewall baseline with SSH preserved
- fail2ban installed/configured, intentionally inactive during ingress/TLS bootstrap
- unattended security upgrades ensured
- optional Caddy ingress using the official Caddy apt repository
- global Caddyfile importing app fragments from `/etc/caddy/sites.d/*.caddy`

Validated on Ubuntu Server through deploy, harden, reboot, redeploy, and smoke checks.

Installed `srvbox` release trees are owned by `root:root`; `srvbox` is infrastructure code executed as root.

## Current repo shape

In [ ]:
%%bash
pwd && tree -aI '.git|__pycache__|.ipynb_checkpoints|CRAFT*|META*|git*|scratch*' --dirsfirst

/app/data/1iis/srvbox


.
├── docs
│   └── mvp-roadmap.md
├── [01;34

mhost
│   ├── sync.ipynb
│   └── sync.py
├── [01;34

mscripts
│   ├── README.md
│   ├── check.sh
│   ├

── deploy.sh
│   └── setup.sh
├── templates


│   ├── env.sh.example
│   └── sshd_config.example
├── 

.gitignore
├── LICENSE
└── README.md

5 directories, 12 files


## Implemented files

### `scripts/README.md`

In [ ]:
Markdown(view('scripts/README.md', view_range=(1,96)))  # excerpt, rest is redundant with CRAFT contents

<div class="prose" markdown="1">

# `scripts/`

> Workstation-to-server deployment and remote bootstrapping.

## Philosophy
> Why artifact deployment?

A production server should not need `git` or GitHub credentials, let alone write access to source repositories.  
Instead, the workstation sends a complete artifact `/tmp/deploy/1iis/srvbox.tar.gz` which the server unpacks and installs.  
This gives us a small, inspectable deployment path: `tar`, `cat`, `scp`, `ssh`, `sudo`, `sh`, with two clear network ops (🛜 below), in three `.sh` scripts:
1. [`deploy.sh`](#1-deploysh)
2. [`setup.sh`](#2-setupsh)
3. [`check.sh`](#3-checksh)

In practice, you only need to run the first one from your workstation.

```sh
./scripts/deploy.sh -d user@host
```
does:

```text
——— Workstation ———————————————————————————————————————————————————
 1    tar repo                   /tmp/deploy/1iis/srvbox.tar.gz
 2    cat > runner script        /tmp/deploy/1iis/srvbox.run.sh
       🠋
 3    scp tarball + runner       🛜 upload both files to server
 4    ssh -t "sudo sh runner"    🛜 elevated shell script
       🠋
——— Server (as root) ——————————————————————————————————————————————
 5    runner extracts tarball    sees scripts/setup.sh
       🠋
··· /tmp/deploy/1iis/srvbox/
 6    setup.sh                   prod dir copy + link to "current"
       🠋
··· /opt/1iis/srvbox/releases/<stamp>/ → /opt/1iis/srvbox/current/
 7    sync.py apply              converge host to desired state
       🠋
——— Hardened host —————————————————————————————————————————————————
```

---

## 1. `deploy.sh`
> Workstation side

```text
./scripts/deploy.sh -s /path/to/repo -d user@host
```

1. Two files:
    - `$REPO.tar.gz`: clean tarball of the repo (excludes `.git`, `__pycache__`, temporary files).
    - `$REPO.run.sh`: temporary runner script.
2. `scp` uploads both in a single transaction to `/tmp/deploy/1iis/$REPO/`
3. `ssh -t` executes the runner remotely, so `sudo` has a TTY for the password prompt.


### `srvbox.run.sh`
> Temporary runner script

#### SSH and `sudo` TTY trap

To execute our setup on a remote host, we need to run an installation script with `sudo`. The naive approach is to use a here-doc over SSH:

```bash
# ANTI-PATTERN: Do not do this
ssh -t user@host "sudo sh -s" <<'EOF'
  echo "Doing root things"
  ./scripts/setup.sh
EOF
```

This fails disastrously if the remote user requires a password for `sudo`.

When `ssh -t` allocates a pseudo-terminal (TTY), `sudo` uses it to prompt for the user's password. However, because we are piping the script into `stdin` via the here-doc, `sudo` reads **from that same `stdin` stream**. It consumes the literal text of the deployment script and interprets it as the user's password attempt, printing it in cleartext to the terminal (which may leak into logs) and obviously fails authentication.

We also cannot rely on caching `sudo` credentials across multiple SSH calls, that cache is tied to the specific TTY session.

#### Generated Runner

To safely elevate privileges without `stdin` collisions, the execution environment (TTY) and the script payload (file content) must be separated. We generate a tiny "runner" script locally, upload it alongside the tarball artifact, and execute it over a single interactive SSH session:

```bash
# 1. Generate runner locally
write_runner "$runner_file"

# 2. Upload both artifact and runner
scp "$artifact" "$runner_file" "$DST:/tmp/deploy/1iis/"

# 3. Execute interactively
ssh -t "$DST" "sudo env REPO='$repo' sh '/tmp/deploy/1iis/$remote_runner'"
```

By placing the commands in a remote file, `sudo` prompts securely via the TTY, and `sh` reads the commands safely from the filesystem. They do not fight over `stdin`.

`sudo` without a TTY refuses to prompt for a password in some configurations, so `ssh -t` forces pseudo-terminal allocation.

</div>

### `scripts/deploy.sh`

Workstation-side deployment command.

Usage:

```sh
./scripts/deploy.sh -s /path/to/repo -d user@host
```

Defaults:

```text
-s / --src = .
-d / --dst = required SSH destination
DEPLOY_ROOT = /tmp/deploy/1iis
```

It:

1. creates a tarball artifact of the repo
2. excludes dev/cache files
3. normalizes tar ownership with `--owner=0 --group=0 --numeric-owner`
4. writes a temporary remote runner script
5. uploads artifact + runner to the target server
6. runs the runner with `ssh -t ... sudo ...`

In [ ]:
print(view('scripts/deploy.sh'))

#!/usr/bin/env bash
set -euo pipefail

SRC="."
DST=""
DEPLOY_ROOT="${DEPLOY_ROOT:-/tmp/deploy/1iis}"

usage() {
  cat <<'EOF'
Usage: scripts/deploy.sh [-s SRC] -d USER@HOST

Build an artifact from SRC, upload it to USER@HOST, extract it under
/tmp/deploy/1iis/REPO, then run the repo's scripts/setup.sh remotely.

Options:
  -s, --src   Source repository path. Defaults to current directory.
  -d, --dst   SSH destination, e.g. onei@1.2.3.4. Required.
  -h, --help  Show this help.
EOF
}

log() {
  printf '==> %s\n' "$*"
}

fail() {
  echo "error: $*" >&2
  exit 1
}

parse_args() {
  while [ "$#" -gt 0 ]; do
    case "$1" in
      -s|--src)
        [ "$#" -ge 2 ] || fail "$1 requires a value"
        SRC="$2"
        shift 2
        ;;
      -d|--dst)
        [ "$#" -ge 2 ] || fail "$1 requires a value"
        DST="$2"
        shift 2
        ;;
      -h|--help)
        usage
        exit 0
        ;;
      *)
        fail "unknown argument: $1"
        ;;
    esac
  done

  [ -n "$DST" ] 

### `scripts/setup.sh`

Server-side installer.

It expects to be run as root from an unpacked artifact.

It:

1. installs base packages:
   ```text
   python3 ca-certificates
   ```
2. copies the artifact into:
   ```text
   /opt/1iis/srvbox/releases/<UTC timestamp>/
   ```
3. normalizes installed release ownership/perms:
   ```text
   root:root
   ```
4. updates:
   ```text
   /opt/1iis/srvbox/current
   ```
5. runs:
   ```sh
   host/sync.py apply
   ```

In [ ]:
print(view('scripts/setup.sh'))

#!/usr/bin/env bash
set -euo pipefail

APP="${APP:-srvbox}"
BASE="${BASE:-/opt/1iis}"
APP_ROOT="${APP_ROOT:-$BASE/$APP}"
RELEASES="$APP_ROOT/releases"
CURRENT="$APP_ROOT/current"
PYTHON="${PYTHON:-python3}"

log() {
  printf '==> %s\n' "$*"
}

fail() {
  echo "error: $*" >&2
  exit 1
}

need_root() {
  if [ "$(id -u)" -ne 0 ]; then
    fail "setup.sh must run as root"
  fi
}

source_dir() {
  cd "$(dirname "$0")/.."
  pwd -P
}

install_base_packages() {
  log "Ensuring base packages are installed"

  if command -v apt-get >/dev/null 2>&1; then
    apt-get update
    DEBIAN_FRONTEND=noninteractive apt-get install -y python3 ca-certificates
  else
    fail "only apt-based systems are supported for now"
  fi
}

copy_release() {
  local src="$1"
  local stamp
  stamp="$(date -u +%Y%m%dT%H%M%SZ)"
  local release="$RELEASES/$stamp"
  local tmp="$release.tmp"

  log "Installing $APP release: $release"
  mkdir -p "$RELEASES"
  chown root:root "$APP_ROOT" "$RELEASES"
  chmod 0755 "$APP_ROOT" "$

### `scripts/check.sh`

Smoke checks:

```text
bash -n scripts/*.sh
python3 -m py_compile host/sync.py
python3 host/sync.py status
```

In [ ]:
print(view('scripts/check.sh'))

#!/usr/bin/env bash
set -euo pipefail

log() {
  printf '==> %s\n' "$*"
}

ROOT="$(cd "$(dirname "$0")/.." && pwd -P)"

cd "$ROOT"

log "Checking shell scripts"
for script in scripts/*.sh; do
  bash -n "$script"
done

log "Checking Python syntax"
python3 -m py_compile host/sync.py

log "Checking sync status command"
python3 host/sync.py status

log "All checks passed"


### `host/sync.py`

Stdlib-only Python reconciler exported from the notebook [`host/sync`](dialog_?name=1iis%2Fsrvbox%2Fhost%2Fsync).

CLI:

```text
status
apply
```

`status` is read-only and works without root. It reports host facts for the baseline: OS, root state, SSH, firewall, fail2ban, unattended upgrades, and Caddy.

`apply` requires root and converges the host through explicit idempotent steps:

```text
base directories
access policy
ssh hardening
firewall
fail2ban
unattended upgrades
caddy
```

Key behavior:

- base `/opt/1iis`, `/etc/1iis`, `/var/lib/1iis`, `/var/log/1iis` dirs are ensured as `root:root 0755`
- human users are unmanaged; deploy access must already exist via SSH + sudo
- no `srvbox` daemon/service user exists
- SSH hardening is written as `/etc/ssh/sshd_config.d/90-1iis-srvbox.conf` and validated with `sshd -t`
- `ufw` denies incoming, allows outgoing, and allows SSH before enabling
- fail2ban is installed/configured but disabled by policy during ingress/TLS bootstrap
- unattended upgrades are installed/configured with apt timers enabled
- Caddy ingress is policy-controlled by `ENABLE_CADDY`
- Caddy uses the official Caddy apt repository
- app-owned Caddy fragments live in `/etc/caddy/sites.d/*.caddy`
- apt operations retry to tolerate package-manager lock races

<details><summary>⏩ Tool call</summary>

```python
await view_dlg(dname='host/sync')
```

</details>

In [ ]:
await view_dlg(dname='host/sync')

'<msgs><note id="_5306d87c"># ` sync `\n> `1iis/srvbox/host/sync.ipynb` → `sync.py` source</note><note id="_83041afd">## Imports and constants</note><note id="_29ce56c1">[`sync.py`](file_editor_?fp=%2Fapp%2Fdata%2F1iis%2Fsrvbox%2Fhost%2Fsync.py) starts with stdlib-only imports and a repository root constant. The MVP should stay dependency-free so a fresh Ubuntu/Debian host can run it as soon as `python3` exists.</note><code id="_2dc2a0b2" export>from __future__ import annotations\nimport argparse\nimport os\nimport platform\nimport pwd\nimport shutil\nimport subprocess\nimport sys\nimport time\nfrom dataclasses import dataclass\nfrom pathlib import Path\nfrom typing import Callable</code><code id="_bcdc3460" export>ROOT = Path(__file__).resolve().parents[1]</code><note id="_88fb5a1d">## Small logging and command helpers</note><note id="_abb96096">Keep process helpers tiny and explicit. `log` gives predictable operator output; `run` centralizes subprocess execution so later reconciliati

In [ ]:
print(view('host/sync.py'))

from __future__ import annotations
import argparse
import os
import platform
import pwd
import shutil
import subprocess
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Callable
ROOT = Path(__file__).resolve().parents[1]
def log(msg: str) -> None:
    print(f"==> {msg}")
def run(cmd: list[str], *, check: bool = True) -> subprocess.CompletedProcess[str]:
    log("$ " + " ".join(cmd))
    return subprocess.run(cmd, check=check, text=True)
def require_root() -> None:
    if os.geteuid() != 0:
        raise SystemExit("error: sync.py must run as root")
def require_ubuntuish() -> None:
    if not Path("/etc/os-release").exists():
        raise SystemExit("error: /etc/os-release not found")

    data = Path("/etc/os-release").read_text()
    if "Ubuntu" not in data and "Debian" not in data:
        raise SystemExit("error: only Ubuntu/Debian-like hosts are supported for now")
@dataclass(frozen=True)
class Step:
    name: str
    fn: Callabl

## Key design decisions

### No `git` on production servers

Production servers do not need repository access, GitHub credentials, deploy keys, or even `git`.

Deployments are artifact-based:

```text
tar + scp + ssh + sudo + sh
```

This keeps the production surface small.

### Use uploaded runner file, not SSH here-doc into sudo

Do **not** do this:

```sh
ssh -t user@host "sudo sh -s" <<'EOF'
...
EOF
```

If `sudo` prompts for a password, it may consume the script body as password input.

Correct pattern:

```text
1. upload artifact
2. upload generated runner script
3. ssh -t user@host "sudo sh /tmp/deploy/1iis/repo.run.sh"
```

This separates:

```text
TTY/password input
```

from:

```text
script payload
```

### Root owns infrastructure code

`srvbox` release code under `/opt` is root-owned because `srvbox` is run as root.

Human deploy users may trigger deployment via `sudo`, but they should not own installed production code.

### No `srvbox` service user yet

`srvbox` is currently a reconciler, not a daemon.

If we later add a long-running `srvbox-agent`, create a system user then.

### Future app identity model

Future apps should use per-app system users:

```text
/opt/1iis/<app>/releases/...   root:root
/opt/1iis/<app>/current        root-managed symlink
/etc/1iis/<app>/               root-owned config/secrets
/var/lib/1iis/<app>/           app-user:app-user
/var/log/1iis/<app>/           app-user:app-user
```

systemd services should run as:

```ini
User=iis-<app>
Group=iis-<app>
WorkingDirectory=/opt/1iis/<app>/current
EnvironmentFile=/etc/1iis/<app>/env
```

The app should be able to read its deployed code, but not modify it.

### Caddy as optional host ingress

Caddy belongs to `srvbox` as an optional-but-first-class host capability, not as app/domain logic.

`ENABLE_CADDY = True` means `srvbox` owns the baseline HTTP ingress substrate. `ENABLE_CADDY = False` means public HTTP ingress is closed by policy.

Apps own routing fragments under `/etc/caddy/sites.d/*.caddy`; higher infra owns DNS, VPS provisioning, and secret materialization.

## Roadmap

### Phase 1 — Artifact deployment spine ✅

> [Summary](dialog_?name=1iis/srvbox/META#_4ac57f18): Artifact deployment, release/current structure, root-owned installed releases, base reconciler skeleton.

<details><summary>⏩ Tool call</summary>

```python
await view_msg(dname='META.ipynb', id='_4ac57f18', nums=False)
```

</details>

### Phase 2 — Host hardening ✅

Completed hardening baseline:

- access/identity policy: human users unmanaged, no `srvbox` service user
- SSH hardening by validated drop-in
- firewall baseline with `ufw`
- fail2ban installed/configured, disabled by explicit policy for Caddy/TLS bootstrap
- unattended security upgrades ensured

Validated through deploy, reconnect, reboot, redeploy, and deployed smoke checks.

Completion criteria reached:

```text
fresh Ubuntu VM
→ deploy artifact
→ install root-owned release under /opt/1iis/srvbox/releases/<timestamp>/
→ host/sync.py apply
→ SSH hardened
→ ufw active with SSH allowed
→ fail2ban configured by policy
→ unattended upgrades enabled
→ reboot survives
→ redeploy remains idempotent
```

### Phase 3 — Caddy ingress ✅

Caddy is the optional-but-first-class HTTP ingress layer.

When `ENABLE_CADDY = True`, `srvbox`:

- installs Caddy from the official Caddy apt repository
- owns `/etc/caddy/Caddyfile`
- creates `/etc/caddy/sites.d/`
- imports app-owned fragments from `/etc/caddy/sites.d/*.caddy`
- validates Caddy config before service changes
- opens `80/tcp` and `443/tcp` in `ufw`
- enables/starts or reloads `caddy.service`

When `ENABLE_CADDY = False`, public HTTP ingress is closed by policy.

Caddy deliberately contains no app/domain intent. Apps deploy their own fragments; higher infra may provision DNS/VPS/secrets and call the same deploy primitive.

Contract:

```text
srvbox owns host ingress substrate
apps own routing intent
infra owns orchestration
```

This keeps standalone deploys and future infra automation on the same path.

### Next chapter — App deployment and infra integration

`srvbox` should support three operating modes with the same underlying contracts:

1. **Standalone host** — a repo deploy can harden a VPS and run an app with only SSH, sudo, DNS pointed at the host, and app-provided config.
2. **Infra-orchestrated host** — higher infra provisions VPS/DNS/secrets/storage, then calls the same `srvbox` deploy primitive.
3. **Private/internal host** — services may run behind WireGuard/Tailnet/private networking with no public HTTP ingress.

The important invariant: manual and automated paths should be the same path. Infra adds orchestration around `srvbox`; it should not replace the local deploy/reconcile primitive.

Expected app-layer shape:

- first app deploys copy the `srvbox` artifact/release pattern manually
- abstraction is extracted only after repetition hurts
- app code releases remain root-owned under `/opt/1iis/<app>/releases/`
- `/opt/1iis/<app>/current` is a root-managed symlink
- per-app services run as system users named `iis-<app>`
- app config/secrets live under `/etc/1iis/<app>/`
- app state lives under `/var/lib/1iis/<app>/`
- app logs live under `/var/log/1iis/<app>/`
- systemd units point at the app's current release and use an environment file
- public HTTP apps write Caddy fragments under `/etc/caddy/sites.d/*.caddy`
- private/internal services may skip Caddy entirely

The first concrete app target is the standalone `1i.is` website deployment.